### PR TITLE
Milestone II+ review comments

### PR DESCRIPTION
**Pull request by Wiezzel**
## General remarks
* Update substrate v0.9.20 -> v0.9.28.
* Top-level README is a bit too brief. It would help (e.g. reviewers) if it had the repo structure described.

### CODE REVIEW

File: pallets/ajuna-board/src/lib.rs [hunk 1]
```diff
@@ -157,10 +159,15 @@ pub mod pallet {
...
 			// As this would be ran in L2 we may want to check that we are in L2??
+			//
+			// One way to solve it would be not "adding" a player to a game but "inviting" them.
+			// After being invited, a player (AccountId) would be able to accept or decline it.
 			let _ = ensure_signed(origin)?;
 
 			// Ensure we have players
 			ensure!(!players.is_empty(), Error::<T>::NotEnoughPlayers);
+			// Maybe using a global board counter here, incremented (+1) on every new_game().
+			// That way we'd avoid game creators being aware of available "board_ids".
```
Comment (cowboy-bebug) [inline pallets/ajuna-board/src/lib.rs:170]: Yeah good point.
We probably need to change `BoardStates` to be a double map of `GameCreator, BoardId => BoardState` so the game creators are aware only of what they created.

File: pallets/ajuna-board/src/lib.rs [hunk 2]
```diff
@@ -224,6 +238,12 @@ pub mod pallet {
 		#[pallet::weight(T::WeightInfo::finish_game())]
 		pub fn finish_game(origin: OriginFor<T>, board_id: T::BoardId) -> DispatchResult {
 			// TODO if this is L2 do we really need to check the origin?
+			//
+			// There is no check for the "origin", so is anyone able to finish any game?
+			// What if the game is still running (no winner)?
+			// Maybe we should allow the game's creator and its players to finish it.
+			// Or finish it right after "declare_winner()" is called, as we already deposit an
+			// "Event::GameFinished" there.
```
Comment (cowboy-bebug) [inline pallets/ajuna-board/src/lib.rs:246]: Yeah we need to keep track of which origin created games so only they can finish games (related to the comment above).

Just sharing more context about this pallet - all the extrinsics in this pallet should be signed and checked by TEE enclave key, which isn't supported yet. And this pallet belongs to L2 runtime under TEE.

File: pallets/ajuna-gameregistry/src/lib.rs [hunk 3]
```diff
@@ -48,6 +48,7 @@ pub mod pallet {
 	};
 	use frame_system::pallet_prelude::*;
 
+	// Vec<T> -> BoundedVec<AccountId, MaxAccounts>
```
Comment (cowboy-bebug) [inline pallets/ajuna-gameregistry/src/lib.rs:51]: :+1: again we'll check with Integritee whether `BoundedVec` is supported.

File: pallets/ajuna-gameregistry/src/lib.rs [hunk 4]
```diff
@@ -102,6 +103,7 @@ pub mod pallet {
 	#[pallet::without_storage_info]
 	pub struct Pallet<T>(_);
 
+	// Maybe this pallet is still in an early stage of development, but, why not use Events?
```
Comment (cowboy-bebug) [inline pallets/ajuna-gameregistry/src/lib.rs:106]: Yes, this pallet is still actively being developed for L2.
We were notified that Events weren't supported as a callback mechanism from L2 yet. We're relying on storage persistence to communicate between L1 and L2 at the moment as an interim solution.

File: runtime/bajun/src/lib.rs [hunk 5]
```diff
@@ -22,6 +22,9 @@
 #[cfg(feature = "std")]
 include!(concat!(env!("OUT_DIR"), "/wasm_binary.rs"));
 
+// Review comment: A lot o this code is copied from runtime/ajuna.
+// Maybe this duplication could be avoided by using feature toggles?
+
```
Comment (cowboy-bebug) [inline runtime/bajun/src/lib.rs:27]: Yeah we'll factor duplication out into `runtime/common` or something like that. Thanks for spotting this.

**Review by cowboy-bebug — COMMENTED** [review]
Thanks for the review. I left some comments to provide more context.

File: ajuna-common/src/lib.rs [hunk 6]
```diff
@@ -61,6 +61,8 @@ pub trait GetIdentifier<T: Identifier> {
 	fn get_identifier() -> T;
 }
 
+// Review comment: Why not use `BoundedVec` for `State`?
+// There's `MAX_STATE_LENGTH` defined later and manually implemented `MaxEncodedLen`.
```
[the remark this answers is not in the document]
Comment (cowboy-bebug) [reply ajuna-common/src/lib.rs:65]: Confirmed that `BoundedVec` is supported so will update accordingly.

File: pallets/ajuna-gameregistry/src/lib.rs (hunk 3 above)
Comment (cowboy-bebug) [reply pallets/ajuna-gameregistry/src/lib.rs:51]: We'll update this.

File: pallets/ajuna-board/src/lib.rs (hunk 2 above)
Comment (andyjsbell) [reply pallets/ajuna-board/src/lib.rs:246]: Just to add, this pallet is executed in a Trusted Execution Environment (TEE) and probably should be pulled out into its own repo to make this clear as the pallet and its environment aren't strictly a substrate runtime.  In other words there are small workarounds for this sidechain.  Having said all of that your point is valid and we intend to, as @cowboy-bebug has pointed out, to put in a guard for this at a later date.

**Comment by darkfriend77** [comment]
Closing this one because most of the parts have been implemented, through ongoing development.

**Pull request closed without merging**